content/blog/2024-10-21-testing-the-limits/index.md

```diff
@@ -0,0 +1,239 @@
+- `-P` - this specifies that we would like the tests to use 16 pipelines (send 16 operations per request). 
+- `-q` - silences the output to show only the final results
+- `-a` - use the specified password
+- `-h` - run the test against the specified host
+
+Honestly, I was asonished by the first set of results I got. Sure, I expected Valkey to be fast but this speed from a single board computer?? ON A SINGLE THREAD?? Its amazing. 
+
+```bash
+redis-benchmark -n 1000000 -t set,get -P 16 -q -a e41fb9818502071d592b36b99f63003019861dad  -h 10.0.1.136
+SET: 173040.33 requests per second, p50=4.503 msec                    
+GET: 307031.00 requests per second, p50=2.455 msec
+```
+
+Between the two tests we averaged 240,000 requests per second with a median time of 3.45 msec (or 0.0034 seconds) for each operation. 
+
+## Raising the CPU clock speed
+
+Since Valkey is a single threaded application it makes sense that higher clock speeds would lead to more performance. I don't expect most people will overclock their servers in production. But different servers may be available with different CPU clock speeds. 
+
+**Note:** Clock speeds generally are only comparable between CPU's with a similar architecture. For example, you could reasonably compare clock speeds between an 12th generation Intel i5 and a 12th generation Intel i7. If the 12th gen i7 had a max clock speed of 5Ghz that doesnt necessarily mean it will be slower than a AMD Ryzen 9 9900X clocked at 5.6Ghz. 
+
+If you're following along on a Pi of your own I've outlined the steps to overclock your CM4 below. Otherwise you can skip to the results section below.
+
+1. Open the below file in a text editor like nano.
+    - `sudo nano /boot/firmware/config.txt`
+2. At the end of the file add this section below the `[all]` tag.
+    ```
+    [all]
+    over_voltage=8
+    arm_freq=2200
+    ```
+3. Save with cmd + x, y, enter 
+4. Restart the server and log back in `sudo restart now`
+
+We've just increased the speed of the Pi by 47% by raising the clock speed from 1.5Ghz to 2.2Ghz. Now lets re-run our test and see how things look!
+
+```bash 
+redis-benchmark -n 1000000 -t set,get -P 16 -q -a e41fb9818502071d592b36b99f63003019861dad -h 10.0.1.136
+SET: 394368.41 requests per second, p50=1.223 msec                    
+GET: 438058.53 requests per second, p50=1.135 msec 
+```
+
+We're up to 416,000 requests per second with a 1.2 msec (0.0012s) median latency. The mathmaticians out there might notice that this speed up is a lot more than the expected 47% increase. It's a 73% increase in requests per second and a 65% improvement in processing time. Whats happening?!
+
+While I haven't gone to machine code to verify here is what I believe is happening. The memory controller for the CM4 also lives on the CPU. Increasing the clock speed of the CPU also increases the performance of the memory. This means we almost are doubling the expected speedup by increasing the clock speed.
```

Comment:
Interesting observation, the GET time is not too far off from the theoretical estimation (307k * 1.47 = 451k). While the set is significantly faster. I'm not sure why that would be though.

Reply:
Yeah it seems to be the set operations that had a big increase which effected the overall change. Do you think it would be better to put the results in a table like: 

| Operation | Actions | Change | 
| ---------- | -------- | --------|
| SET | 394k | +127% |
| GET | 438k | +44% |